components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServiceURL.java

```diff
@@ -0,0 +1,276 @@
+    private String hostName;
+    private int port;
+    private String urlPath;
+    private Map<String, String> parameters;
+    private String fragment;
+    private Map<String, String> fragmentParams;
+    private String absoluteUrl;
+    private String relativeUrl;
+
+    public ServiceURL(String protocol, String hostName, int port, String urlPath, Map<String, String> parameters,
+                      String fragment, Map<String, String> fragmentParams) throws URLBuilderException {
+
+        this.protocol = protocol;
+        this.hostName = hostName;
+        this.port = port;
+        this.urlPath = urlPath;
+        this.parameters = parameters;
+        this.fragment = fragment;
+        this.fragmentParams = fragmentParams;
+        this.absoluteUrl = getAbsoluteUrl();
+        this.relativeUrl = getRelativeUrl();
+    }
+
+    /**
+     * This method is called to get the protocol of {@link ServiceURL}.
+     *
+     * @return String of the protocol.
+     */
+    public String getProtocol() {
+
+        return protocol;
+    }
+
+    /**
+     * This method is called to get the host name of {@link ServiceURL}.
+     *
+     * @return String of the host name.
+     */
+    public String getHostName() {
+
+        if (hostName.endsWith("/")) {
+            hostName = hostName.substring(0, hostName.length() - 1);
+        }
+        return hostName;
+    }
+
+    /**
+     * This method is called to get the port of {@link ServiceURL}.
+     *
+     * @return value of the port.
+     */
+    public int getPort() {
+
+        return port;
+    }
+
+    /**
+     * This method is called to get the Url path of {@link ServiceURL}.
+     *
+     * @return String of the url path.
+     */
+    public String getUrlPath() {
+
+        return urlPath;
+    }
+
+    /**
+     * This method is called to get the parameter value when the key is provided.
+     *
+     * @param key Key of the parameter.
+     * @return The value of the parameter.
+     */
+    public String getParameter(String key) {
+
+        return parameters.get(key);
+    }
+
+    /**
+     * This method is called to get a list of the parameter names.
+     *
+     * @return ArrayList of parameter keys.
+     */
+    public ArrayList<String> getParameterKeys() {
+
+        ArrayList<String> parameterArrayList = new ArrayList<>();
+
+        for (Map.Entry<String, String> entry : this.parameters.entrySet()) {
+            parameterArrayList.add(entry.getKey());
+        }
+        return parameterArrayList;
+    }
+
+    /**
+     * This method is called to get the decoded fragment from the url.
+     *
+     * @return The decoded fragment.
+     * @throws URLBuilderException
+     */
+    public String getFragment() throws URLBuilderException {
+
+        URI uri;
+        try {
+            uri = new URI(this.getAbsoluteUrl());
+        } catch (URISyntaxException e) {
+            throw new URLBuilderException("Error while trying to build the url", e);
+        }
```

Comment:
Why build a URI. Why not return from fragment variable

Reply:
Went this way because the fragments are set in two ways. Either by setFragment(String fragment) where the fragment is set via a String, or by setFragmentParameters(String key, String value).. When the fragment is added via the second method, we don't store the fragment as a string before we encode it and append to the url. To resolve this, we can introduce a variable to store the fragment as a string before it is encoded and appended to the URL.

Reply:
setFragement denotes -> fragement is fully replaced
addFragementParameter -> denotes key value par is added as to the fragment and you can call it many times to add.